File: packages/js-sdk/src/backends/openai-compatible.ts
import { BadRequestError } from '../errors.js';

export interface GatewayEndpoints {
  /** Origin plus `/v1`, with no trailing slash. Chat, audio, and models live here. */
  apiBase: string;
  /** `GET /health` is on the gateway root, not under `/v1`. */
  healthUrl: string;
}

/**
 * Accept either form from `.env.example` and from the OpenAI client habit:
 *
 * - `http://localhost:8080`
 * - `http://localhost:8080/v1`
 *
 * `/health` is never under `/v1`. A trailing slash is ignored.
 */
export function resolveOpenAICompatibleBase(baseURL: string): GatewayEndpoints {
  let url: URL;
  try {
    url = new URL(baseURL);
  } catch (cause) {
    throw new BadRequestError(`baseURL is not a valid URL: ${baseURL}`, { cause });
  }
  if (url.protocol !== 'http:' && url.protocol !== 'https:') {
    throw new BadRequestError('baseURL must use http or https.');
  }

  const trimmed = baseURL.trim().replace(/\/+$/u, '');
  const apiBase = trimmed.endsWith('/v1') ? trimmed : `${trimmed}/v1`;
  const root = apiBase.slice(0, -'/v1'.length);
  return { apiBase, healthUrl: `${root}/health` };
}

File: packages/python-sdk/src/natlas/backends/openai_compatible.py
"""URL rules for the ``/serve`` gateway and any OpenAI-compatible N-ATLaS endpoint."""

from __future__ import annotations

from urllib.parse import urlsplit

from natlas.errors import BadRequestError


def resolve_openai_compatible_base(base_url: str) -> tuple[str, str]:
    """Return ``(api_base, health_url)``.

    Both of these are accepted:

    * ``http://localhost:8080``
    * ``http://localhost:8080/v1``

    ``/health`` is never under ``/v1``. A trailing slash is ignored.
    """
    parsed = urlsplit(base_url.strip())
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        raise BadRequestError(f"base_url must be an http(s) URL, got {base_url!r}.")
    trimmed = base_url.strip().rstrip("/")
    api_base = trimmed if trimmed.endswith("/v1") else f"{trimmed}/v1"
    root = api_base[: -len("/v1")]
    return api_base, f"{root}/health"
